patsy/util.py

```diff
@@ -550,3 +562,127 @@ def test_iterable():
     assert iterable({"a": 1})
     assert not iterable(1)
     assert not iterable(iterable)
+
+##### Handling Pandas's categorical stuff is horrible and hateful
+
+# Basically they decided that they didn't like how numpy does things, so their
+# categorical stuff is *kinda* like how numpy would do it (e.g. they have a
+# special ".dtype" attribute to mark categorical data), so by default you'll
+# find yourself using the same code paths to handle pandas categorical data
+# and other non-categorical data. BUT, all the idioms for detecting
+# categorical data blow up with errors if you try them with real numpy dtypes,
+# and all numpy's idioms for detecting non-categorical types blow up with
+# errors if you try them with pandas categorical stuff. So basically they have
+# just poisoned all code that touches dtypes; the old numpy stuff is unsafe,
+# and you must use special code like below.
+#
+# Also there are hoops to jump through to handle both the old style
+# (Categorical objects) and new-style (Series with dtype="category").
+
+# Needed to support pandas < 0.15
+def pandas_Categorical_from_codes(codes, categories):
+    assert have_pandas_categorical
+
+    # Old versions of pandas sometimes fail to coerce this to an array and
+    # just return it directly from .labels (?!).
+    codes = np.asarray(codes)
+    if hasattr(pandas.Categorical, "from_codes"):
+        return pandas.Categorical.from_codes(codes, categories)
+    else:
+        return pandas.Categorical(codes, categories)
+
+def test_pandas_Categorical_from_codes():
+    c = pandas_Categorical_from_codes([1, 1, 0, -1], ["a", "b"])
+    assert np.all(np.asarray(c)[:-1] == ["b", "b", "a"])
+    assert np.isnan(np.asarray(c)[-1])
+
+# Needed to support pandas < 0.15
+def pandas_Categorical_categories(cat):
+    # In 0.15+, a categorical Series has a .cat attribute which is a
+    # Categorical object, and Categorical objects are what have .categories /
+    # .codes attributes.
+    if hasattr(cat, "cat"):
```

Comment:
Oh good. Thanks for the update.

On Wed, Apr 15, 2015 at 5:42 PM, Jan Schulz <notifications@github.com>
wrote:

> In patsy/util.py
> <https://github.com/pydata/patsy/pull/59#discussion_r28466016>:
>
> > +    if hasattr(pandas.Categorical, "from_codes"):
> > +        return pandas.Categorical.from_codes(codes, categories)
> > +    else:
> > +        return pandas.Categorical(codes, categories)
> > +
> > +def test_pandas_Categorical_from_codes():
> > +    c = pandas_Categorical_from_codes([1, 1, 0, -1], ["a", "b"])
> > +    assert np.all(np.asarray(c)[:-1] == ["b", "b", "a"])
> > +    assert np.isnan(np.asarray(c)[-1])
> > +
> > +# Needed to support pandas < 0.15
> > +def pandas_Categorical_categories(cat):
> > +    # In 0.15+, a categorical Series has a .cat attribute which is a
> > +    # Categorical object, and Categorical objects are what have .categories /
> > +    # .codes attributes.
> > +    if hasattr(cat, "cat"):
>
> hasattr(data, "cat") has become more or less API for identifying a Series
> of dtype category since there was a bugreport, which made that py3
> compatible, so you are right to use that...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pydata/patsy/pull/59/files#r28466016>.
>



-- 
Nathaniel J. Smith -- http://vorpus.org